# HG changeset patch
# Parent 51e561958f11d50aacf4703dfe4c1806ded04a00

Allow blacklisting of product number 1 (legacy PV drivers) when the VM
has the new device id.

Signed-off-by: Paul Durrant <paul.durrant@citrix.com>

diff --git a/xenstore.c b/xenstore.c
index 0682ffe..3497537 100644
--- a/xenstore.c
+++ b/xenstore.c
@@ -1024,6 +1024,8 @@ int
 xenstore_pv_driver_build_blacklisted(uint16_t product_nr,
                                      uint32_t build_nr)
 {
+    int blacklist = 0;
+
     /*
      * In qemu-xen-unstable, this is the master registry of product
      * numbers.  If you need a new product number allocating, please
@@ -1043,7 +1045,6 @@ xenstore_pv_driver_build_blacklisted(uint16_t product_nr,
     case 2:
     case 0xffff: {
         char *buf = NULL;
-        char *tmp;
         const char *product = NULL;
 
         switch (product_nr) {
@@ -1052,16 +1053,27 @@ xenstore_pv_driver_build_blacklisted(uint16_t product_nr,
         case 0xffff: product = "experimental";  break;
         }
 
-        if (product == NULL || asprintf(&buf, "/mh/driver-blacklist/%s/%d", product, build_nr) < 0)
-            return 0;
-        tmp = xs_read(xsh, XBT_NULL, buf, NULL);
-        free(tmp);
-        free(buf);
-        if (tmp == NULL)
-            return 0;
-        else
-            return 1;
+        if (product && asprintf(&buf, "/mh/driver-blacklist/%s/%d", product, build_nr) >= 0) {
+            char *tmp = xs_read(xsh, XBT_NULL, buf, NULL);
+            free(tmp);
+            free(buf);
+            if (tmp != NULL)
+                blacklist = 1;
+        }
+
+        break;
     }
+    default:
+        break;
+    }
+
+    switch (product_nr) {
+    case 1:
+        if (platform_device_id != 0x0001)
+            blacklist = 1;
+
+        break;
+
     case 3:
     case 4: {
         uint16_t device_id;
@@ -1071,19 +1083,19 @@ xenstore_pv_driver_build_blacklisted(uint16_t product_nr,
         revision = build_nr & 0xff;
 
         if (product_nr == 3 && device_id == 0)
-            return 0;
+            break;
 
         if (device_id != platform_device_id ||
             revision < platform_revision)
-            return 1;
-        else
-            return 0;
+            blacklist = 1;
+
+        break;
     }
     default:
-        /* Don't know what product this is -> we can't blacklist
-         * it. */
-        return 0;
+        break;
     }
+
+    return blacklist;
 }
 
 void xenstore_record_dm_state(const char *state)
